07 and 09 post comments and get users

mikegreenan98 opened the pull request.

mikegreenan98 committed:
07 and 09 nearly done

mikegreenan98 committed:
07+09 done - ready for review

mikegreenan98 committed:
07+09 with conflicts of merge resolved

johnhutcheon commented on code/models/models.js:
```diff
@@ -89,3 +89,58 @@ exports.updateArticleVotes = (request) => {
+  INSERT INTO comments
+  (body, author, article_id, votes) 
+  VALUES
+  ('${commentToInsert.body}','${commentToInsert.author}',${commentToInsert.article_id},${commentToInsert.votes})
```
This is SQL injection. Please use the $ syntax to prevent this.

mikegreenan98 replied:
Done - chaged to the $ syntax and added a valueArray into the db.query as 2nd argument

johnhutcheon commented on code/controllers/controllers.js:
```diff
@@ -59,4 +62,61 @@ exports.pushArticleVotes = (req, res, next) => {
+  selectOneArticle(req)
+    .then((result) => {
+      // Secondly, need to check user exists before trying to insert a comment
+      selectOneUser(req)
+        .then((result) => {
+          // now safe to insert comment because user exists
```
SQL will already throw is an error if we try to insert a comment with a user that doesn't exist.

mikegreenan98 replied:
done - have now stripped out my bespoke error handling and put PSQL error handling in

johnhutcheon commented on code/controllers/controllers.js:
```diff
@@ -59,4 +62,61 @@ exports.pushArticleVotes = (req, res, next) => {
+
+exports.postComment = (req, res, next) => {
+  //Firstly, need to check of article exists before trying to insert a comment fo that article
+  selectOneArticle(req)
+    .then((result) => {
```
SQL will throw an error if we try to insert a comment with an article ID that doesn't exist.

mikegreenan98 replied:
sorted via the other changes

johnhutcheon commented on code/controllers/controllers.js:
```diff
@@ -59,4 +62,61 @@ exports.pushArticleVotes = (req, res, next) => {
+            .then((result) => {
+              res.status(200).send({ comment: result });
+            })
+            .catch((err) => {
+              next(err);
+            });
+        })
+        .catch((err) => {
+              next(err);
+        });
+    })
+    .catch((err) => {
+              next(err);
+    });
+};
```
This will prevent errors from actually being passed. This is because you have nested catches. It should be at the most outer point you can reach it.

You only need the code from line 75 to 81 for this to work.

mikegreenan98 replied:
done - have now stripped out my bespoke error handling and put PSQL error handling in, so this whole block of code was reduced to just one call and catch. Also Emily spotted that I did not have 'return' in each of the calls so that is why I needed to have all those extra catch's.

johnhutcheon commented on __tests__/app.test.js:
```diff
@@ -395,3 +394,151 @@ describe("PATCH /api/articles/:article_id --- error handling", () => {
+        expect(data.body.comment[0]).toMatchObject({
+          comment_id: expect.any(Number),
+          body: expect.any(String),
+          article_id: expect.any(Number),
```
We can expect the actual article ID here rather than any number

mikegreenan98 replied:
Done. I have kept line 435 as is, but added this below:
        expect(data.body.comment[0].article_id).toBe(2);
For future reference, not sure if I could do that specifically on line 435 somehow?

johnhutcheon commented on __tests__/app.test.js:
```diff
@@ -395,3 +394,151 @@ describe("PATCH /api/articles/:article_id --- error handling", () => {
+        });
+      });
+  });
+
```
Please can we change these tests to show a 201 status when posting a new comment rather than 200.

mikegreenan98 replied:
Done - now get 201 for all successful POSTs

johnhutcheon commented on __tests__/app.test.js:
```diff
@@ -395,3 +394,151 @@ describe("PATCH /api/articles/:article_id --- error handling", () => {
+describe("POST /api/articles/:article_id/comments - error handling", () => {
+  const validInput = { username: "lurker", body: "a valid body" };
+
+  test("when username provided is not in users DB, return 400 - Bad Request - User does not exist", () => {
+  const badInput = { username: "FREDDIE5000", body: "a valid body" };
+    return request(app)
+      .post("/api/articles/2/comments")
+      .send(badInput)
+      .expect(400)
+      .then((data) => {
+        const expected = { msg: "Bad Request - User does not exist" };
+        expect(data.body).toEqual(expected);
+      });
```
We should expect a 404 in this instance as the username is not found

mikegreenan98 replied:
Done the change

johnhutcheon commented on code/models/models.js:
```diff
@@ -89,3 +89,58 @@ exports.updateArticleVotes = (request) => {
+  });
+};
+
+exports.selectOneUser = (req) => {
+  let user = req.body.username;
+  // if (!Number.isInteger(+user)) {
+  //   return Promise.reject("Bad Request - User does not exist");
+  // }
+  let queryString = `
+  SELECT *
+  FROM users
+  WHERE username = '${user}'`;
+  return db.query(queryString).then((result) => {
+    if (result.rowCount === 0) {
+      return Promise.reject("Bad Request - User does not exist");
+    }
+    return result.rows;
+  });
+};
```
This can now be deleted once you have refactored number 7. We will only need the one model in the controller for posting.

mikegreenan98 replied:
DOne, code refractored.

johnhutcheon reviewed (changes requested):
Hi Mike. Could you please refer to my comments and resubmit. Think there's been a slight misunderstanding for Number 7.

mikegreenan98 committed:
07+09 fixes to address review comments

mikegreenan98 commented:
Hi John
I've addressed all those comments, I resent a PR last evening, but did not
re-create another PR, but you can still see my code changes on the 09...
branch.
Hopefully all OK.
I also not sure how I merge these in - does it merge in my latest changes
or the first submission?
Regards
Mike

On Thu, 23 Feb 2023 at 14:08, johnhutcheon ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> Hi Mike. Could you please refer to my comments and resubmit. Think there's
> been a slight misunderstanding for Number 7.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mikegreenan98/MikeG-BE-News/pull/9#pullrequestreview-1311390662>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A4PD6IDOPCNDNWUVDPOJCULWY5VMTANCNFSM6AAAAAAVFSPZ6U>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Mike Greenan

johnhutcheon reviewed (approved):
Happy for you to merge Mike, thanks for making all those changes! Great work!
